### PR TITLE
DGR-264: Add custom attribute mapping sections

### PR DESCRIPTION
**Pull request by kei178**
This PR adds attribute mapping sections to mod_form to allow users to save course field, course custom filed, and user profile field attribute mappings.

Screenshot 2024-05-07 at 9 27 08

### CODE REVIEW

File: classes/local/attributemapping.php [hunk 1]
```diff
@@ -100,6 +100,8 @@ private function validate_table($table) {
      * @throws InvalidArgumentException If the field name is invalid
      */
     private function validate_field($table, $field) {
+        if (!$field) return;
```
Comment (kei178) [inline classes/local/attributemapping.php:103]: Allowing non-selected values as users could leave the field blank and save the form.

File: classes/local/attributemapping_list.php [hunk 2]
```diff
@@ -81,6 +81,8 @@ public function get_text_content() {
     private function validate_attributemapping($attributemappings) {
         $uniqueattributes = [];
         foreach ($attributemappings as $attributemapping) {
+            if (!$attributemapping->accredibleattribute) continue;
```
Comment (kei178) [inline classes/local/attributemapping_list.php:84]: Exclude non-selected values from uniqueness validation

File: classes/local/attributemapping.php [hunk 3]
```diff
@@ -100,6 +100,10 @@ private function validate_table($table) {
      * @throws InvalidArgumentException If the field name is invalid
      */
     private function validate_field($table, $field) {
+        if (!$field) {
+            return;
+        }
```
Comment (kei178) [inline classes/local/attributemapping.php:105]: Allowing non-selected values as users could leave the field blank and save the form.

File: classes/local/attributemapping.php [hunk 4]
```diff
@@ -116,6 +120,10 @@ private function validate_field($table, $field) {
      * @throws InvalidArgumentException If the ID is invalid
      */
     private function validate_id($table, $id) {
+        if (!$id) {
+            return;
+        }
```
Comment (kei178) [inline classes/local/attributemapping.php:125]: Allowing non-selected values as users could leave the field blank and save the form.

File: classes/local/attributemapping_list.php [hunk 5]
```diff
@@ -81,6 +81,10 @@ public function get_text_content() {
     private function validate_attributemapping($attributemappings) {
         $uniqueattributes = [];
         foreach ($attributemappings as $attributemapping) {
+            if (!$attributemapping->accredibleattribute) {
+                continue;
+            }
```
Comment (kei178) [inline classes/local/attributemapping_list.php:86]: Excluding non-selected values from uniqueness validation.

File: classes/local/formhelper.php [hunk 6]
```diff
@@ -66,4 +68,108 @@ public function load_grade_item_options($courseid) {
...
+    /**
+     * Load course field options for dropdown selection.
+     *
+     * This function retrieves all column names from the 'course' table in the database
+     * and formats them as options for a select element.
+     *
+     * @return array Associative array of course field names suitable for form dropdown.
+     */
+    public function load_course_field_options() {
+        global $DB;
```
Comment (kanishkad) [inline classes/local/formhelper.php:81]: Looks like $DB is not being used here.

**Review by kanishkad — APPROVED** [review]

**Review by albertpaulp — APPROVED** [review]

File: classes/local/attributemapping.php (hunk 3 above)
Comment (Ndibui) [reply classes/local/attributemapping.php:110]: This should be reverted once multiple selection is implemented.

File: classes/local/attributemapping.php (hunk 4 above)
Comment (Ndibui) [reply classes/local/attributemapping.php:128]: This should be reverted once multiple selection is implemented.

File: classes/local/attributemapping_list.php (hunk 5 above)
Comment (Ndibui) [reply classes/local/attributemapping_list.php:86]: This should be reverted once multiple selection is implemented.

File: mod_form.php [hunk 7]
```diff
@@ -293,6 +293,96 @@ public function definition() {
             $mform->addElement('checkbox', 'completionactivities', get_string('completionissuecheckbox', 'accredible'));
         }
 
+        $attributemappingdefaultvalues =
```
Comment (Ndibui) [inline mod_form.php:296]: This will need to be updated to use templates during multiple attributes implementation

**Review by Ndibui — APPROVED** [review]

File: classes/local/attributemapping.php (hunk 3 above)
Comment (kei178) [reply classes/local/attributemapping.php:116]: Yes, that's true as long as it's restricted to select only one of the fields and leave the other blank.

**Pull request merged**